# Overview

In order to answer most of the customer usage and business insights related questions raised in the OpenShift Workload Fingerprinting project, we need to connect two disparate datasets - the insights operator archive and the pyxis database. That is, we want to use pyxis to determine the product name, architecture, vulnerabilities, vendor, etc corresponding to the container image SHA’s in the insights dataset. In a previous [issue](https://github.com/aicoe-aiops/openshift-workload-fingerprinting/issues/11#issuecomment-867776402), we figured out how to do this for a given SHA, by using curl in the terminal. In this notebook, we will try to do this programmatically, and do it for all the SHA’s available in our dataset. We will then store this merged dataset to an s3 bucket and use it for the rest of the analysis in the project going forward. 

# Pre-requisite

In order to fetch the image name (and other details) for the given 'sha' of the image_id, please complete the pre-requisite described below.

1. Follow the [link](https://source.redhat.com/groups/public/ccs-onboarding-program/ccs_onboarding_wiki/setting_up_a_kerberos_ticket_and_red_hat_idm) in order to set-up a kerberos ticket and Red Hat IdM on your machine.

2. Update the `/etc/krb5.conf` on your machine by setting `dns_canonical_hostname` to `false`, as described in the first 'red box' in this [guide](https://docs.engineering.redhat.com/display/HSSP/Pyxis+access+request)

3. Obtain the kerberos ticket by running, `$ kinit <your_kerberos_username>@IPA.REDHAT.COM`


In this notebook, we map the given "sha" of the image_id for the image layer dataset and container dataset provided from the workload data of the insight operator. 

In [165]:
#!pipenv install BeautifulSoup4
#!pipenv install pyarrow
#!pipenv install progressbar
#!pipenv install boto3
#!pipenv install tqdm

# Importing useful packages

In [1]:
import io
import boto3
import requests
import os
import json
import warnings
import pandas as pd

# from bs4 import BeautifulSoup
from requests_kerberos import HTTPKerberosAuth, OPTIONAL
from dotenv import find_dotenv, load_dotenv
from tqdm import tqdm

In [2]:
load_dotenv(find_dotenv())
warnings.filterwarnings("ignore")
pd.set_option("display.max_columns", None)

# Data Collection

In this section, we will fetch from our s3 bucket the containers dataset and the image layers dataset that have been curated from insights operator archives. To learn more about the general content of datasets, please check out the [getting_started](https://github.com/aicoe-aiops/openshift-workload-fingerprinting/blob/master/notebooks/getting_started_notebook.ipynb) notebook. 

In [3]:
# CEPH Bucket variables
s3_endpoint_url = os.getenv("S3_ENDPOINT")
s3_access_key = os.getenv("S3_ACCESS_KEY")
s3_secret_key = os.getenv("S3_SECRET_KEY")
s3_bucket = os.getenv("S3_BUCKET")

# s3 resource to communicate with storage
s3 = boto3.resource(
    "s3",
    endpoint_url=s3_endpoint_url,
    aws_access_key_id=s3_access_key,
    aws_secret_access_key=s3_secret_key,
)

# access the parquet file as an s3 object

obj1 = s3.Object(
    "DH-PLAYPEN", "ccx/fingerprinting/image_layers/date=2021-05-12/2021-05-12.parquet"
)
obj2 = s3.Object(
    "DH-PLAYPEN", "ccx/fingerprinting/containers/date=2021-05-12/2021-05-12.parquet"
)
# download the file into the buffer
buffer1 = io.BytesIO()
obj1.download_fileobj(buffer1)
buffer2 = io.BytesIO()
obj2.download_fileobj(buffer2)

# read the buffer and create the dataframe
image_layers_df = pd.read_parquet(buffer1)
containers_df = pd.read_parquet(buffer2)

In [4]:
image_layers_df.head(3)

cluster_id  \
0  00003d61-9db1-4757-9cd1-84df271daeb9   
1  00003d61-9db1-4757-9cd1-84df271daeb9   
2  00003d61-9db1-4757-9cd1-84df271daeb9   

                                            image_id  \
0  sha256:337c22cabe530213b14965f9ea69a92dbeb5104...   
1  sha256:337c22cabe530213b14965f9ea69a92dbeb5104...   
2  sha256:337c22cabe530213b14965f9ea69a92dbeb5104...   

                                      layer_image_id  layer_image_level  \
0  sha256:9ebb302e1fb002fb643091710dac46f8258781d...                  0   
1  sha256:a74396a32e85c2feeedf76052ed3297859810c8...                  1   
2  sha256:7db62383a7470afbacfc0fab55d5a182e3c5fa2...                  2   

  first_command     first_arg  \
0  icTsn2s_EIax  2v1NneeWoS_9   
1  icTsn2s_EIax  2v1NneeWoS_9   
2  icTsn2s_EIax  2v1NneeWoS_9   

                                        archive_path  
0  archives/compressed/00/00003d61-9db1-4757-9cd1...  
1  archives/compressed/00/00003d61-9db1-4757-9cd1...  
2  archives/compressed/00/00003d61-9db1-4757-9cd1...

In [5]:
containers_df.head(3)

cluster_id     namespace  \
0  00003d61-9db1-4757-9cd1-84df271daeb9  0LiT6ZNtbpYL   
1  00003d61-9db1-4757-9cd1-84df271daeb9  0LiT6ZNtbpYL   
2  00003d61-9db1-4757-9cd1-84df271daeb9  0LiT6ZNtbpYL   

                                               shape  shape_instances  \
0  sha256:3ecf29979b2722bf4a82a5e7a954e8685820720...                1   
1  sha256:3ecf29979b2722bf4a82a5e7a954e8685820720...                1   
2  sha256:542d007d13008cc1be2dbf03601b954c4452947...                1   

                                            image_id first_command  \
0  sha256:f46f210d6023bec16e68340b484a8881ce46d5e...          None   
1  sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf...  n9CdwzVF-cwZ   
2  sha256:a693c315b775c693dc49c19b7f217762676bc28...  b51B0EZ1bw3c   

      first_arg  init_container  \
0  47DEQpj8HBSa           False   
1  RNOaw_AuQeIY           False   
2  ua-xlwwsvdYd           False   

                                        archive_path  
0  archives/compressed/00/00003d61-9db1-4757-9cd1...  
1  archives/compressed/00/00003d61-9db1-4757-9cd1...  
2  archives/compressed/00/00003d61-9db1-4757-9cd1...

## Mapping the SHA's in `image_id` column of Image layers Dataset

First, we try to form a list of unique image_id from the image layer dataset. Using that list, we will be doing the web scraping followed by the formation of the dataframe with image_id and corresponding product name, summary, vendor, version, and other attributes.

In [10]:
# Creating the list of image_id
arr_imageid = image_layers_df.image_id.unique()
list_imageid = arr_imageid.tolist()

In [11]:
# Size of the image_id list
len(list_imageid[:])

256

In [12]:
kerberos_auth = HTTPKerberosAuth(mutual_authentication=OPTIONAL)

In [174]:
dataframe = pd.DataFrame([])

for i in tqdm(range(len(list_imageid))):
    base_url = "https://pyxis.engineering.redhat.com/v1/images?filter=image_id=="
    image_id = list_imageid[i]
    team_url = base_url + str(image_id)
    # print(team_url)
    r = requests.get(team_url, auth=kerberos_auth, verify=False)
    data = json.loads(r.content)
    if len(data["data"]) > 0:
        if len(data["data"][0]["parsed_data"]["labels"]) > 0:
            df = pd.DataFrame(data["data"][0]["parsed_data"]["labels"])
            table = pd.pivot_table(
                df, values="value", aggfunc=lambda x: x, columns="name"
            )
            table["image_id"] = list_imageid[i]
            table = table.set_index("image_id")
            dataframe = dataframe.append(table)

100%|██████████| 256/256 [04:20<00:00,  1.02s/it]


In [175]:
dataframe_image_id = dataframe
dataframe_image_id.head()

License architecture  \
image_id                                                                  
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f...  GPLv2+       x86_64   
sha256:3574d6c1fcc46e1ebd41b7b887b92035ea182131...  GPLv2+       x86_64   
sha256:c588b4653ef037b1300f9692b986cbeac5dfe5e6...  GPLv2+       x86_64   
sha256:7b879d0179b96fe682a186af60836e4091f1fd31...  GPLv2+       x86_64   
sha256:85a79d95924e4a76b1a6ccd974b2c99df349f074...  GPLv2+       x86_64   

                                                                    build-date  \
image_id                                                                         
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f...  2021-04-30T00:31:42.349887   
sha256:3574d6c1fcc46e1ebd41b7b887b92035ea182131...  2021-04-30T00:22:29.539443   
sha256:c588b4653ef037b1300f9692b986cbeac5dfe5e6...  2021-05-04T21:53:04.228477   
sha256:7b879d0179b96fe682a186af60836e4091f1fd31...  2021-04-29T23:09:36.775464   
sha256:85a79d95924e4a76b1a6ccd974b2c99df349f074...  2021-05-08T07:13:14.635470   

                                                                         com.redhat.build-host  \
image_id                                                                                         
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f...  cpt-1004.osbs.prod.upshift.rdu2.redhat.com   
sha256:3574d6c1fcc46e1ebd41b7b887b92035ea182131...  cpt-1008.osbs.prod.upshift.rdu2.redhat.com   
sha256:c588b4653ef037b1300f9692b986cbeac5dfe5e6...  cpt-1005.osbs.prod.upshift.rdu2.redhat.com   
sha256:7b879d0179b96fe682a186af60836e4091f1fd31...  cpt-1008.osbs.prod.upshift.rdu2.redhat.com   
sha256:85a79d95924e4a76b1a6ccd974b2c99df349f074...  cpt-1007.osbs.prod.upshift.rdu2.redhat.com   

                                                                             com.redhat.component  \
image_id                                                                                            
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f...                    ose-cli-artifacts-container   
sha256:3574d6c1fcc46e1ebd41b7b887b92035ea182131...             openshift-enterprise-cli-container   
sha256:c588b4653ef037b1300f9692b986cbeac5dfe5e6...  golang-github-openshift-oauth-proxy-container   
sha256:7b879d0179b96fe682a186af60836e4091f1fd31...  golang-github-openshift-oauth-proxy-container   
sha256:85a79d95924e4a76b1a6ccd974b2c99df349f074...  golang-github-openshift-oauth-proxy-container   

                                                             com.redhat.license_terms  \
image_id                                                                                
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f...  https://www.redhat.com/agreements   
sha256:3574d6c1fcc46e1ebd41b7b887b92035ea182131...  https://www.redhat.com/agreements   
sha256:c588b4653ef037b1300f9692b986cbeac5dfe5e6...  https://www.redhat.com/agreements   
sha256:7b879d0179b96fe682a186af60836e4091f1fd31...  https://www.redhat.com/agreements   
sha256:85a79d95924e4a76b1a6ccd974b2c99df349f074...  https://www.redhat.com/agreements   

                                                                                          description  \
image_id                                                                                                
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f...  OpenShift is a platform for developing, buildi...   
sha256:3574d6c1fcc46e1ebd41b7b887b92035ea182131...  OpenShift is a platform for developing, buildi...   
sha256:c588b4653ef037b1300f9692b986cbeac5dfe5e6...                             OpenShift OAuth Proxy.   
sha256:7b879d0179b96fe682a186af60836e4091f1fd31...                             OpenShift OAuth Proxy.   
sha256:85a79d95924e4a76b1a6ccd974b2c99df349f074...                             OpenShift OAuth Proxy.   

                                                   distribution-scope  \
image_id                                                                
sha256:337c22cabe530213b14965f9ea69a92dbeb5104f..

In [176]:
dataframe_image_id.shape

(111, 40)

We were able to create a dataframe which maps 111 sha's of the image id provided in the image layer dataset out of 256 sha's in the image layer dataset.

In [6]:
# Uploading the mapping dataset in the DH-PLAYPEN bucket
parquet_buffer = io.BytesIO()
dataframe_image_id.to_parquet(parquet_buffer)
s3_obj = s3.Object(
    s3_bucket, "ccx/fingerprinting/image_layers/dataframe_image_id.parquet"
)
status = s3_obj.put(Body=parquet_buffer.getvalue())

'\n# Uploading the mapping dataset in the DH-PLAYPEN bucket\nparquet_buffer = io.BytesIO()\ndataframe_image_id.to_parquet(parquet_buffer)\ns3_obj = s3.Object(\n    s3_bucket, "ccx/fingerprinting/image_layers/dataframe_image_id.parquet"\n)\nstatus = s3_obj.put(Body=parquet_buffer.getvalue())\n'

The corresponding image_id mapped with the product name is saved in the DH-PLAYPEN bucket in the form of dataframe (_dataframe_image_id.parquet_).  

---

# Mapping SHA's from `image_layer_id` column of Image Layers Dataset

In addition to the `image_id` column, the `image_layer_id` column also contains image SHA's. These SHA's correspond to the layers that make up the image in `image_id`. In this section, we try to form a list of unique image_layer_id from the image layer dataset. Using that list, we will be doing the web scraping followed by the formation of the dataframe with image_layer_id and corresponding product name, summary, vendor, and other attributes.

In [13]:
arr_layer_imageid = image_layers_df.layer_image_id.unique()
list_layer_imageid = arr_layer_imageid.tolist()

In [14]:
# Size of the imae_layers list
len(list_layer_imageid)

752

In [15]:
dataframe = pd.DataFrame([])

for i in tqdm(range(len(list_layer_imageid))):
    base_url = "https://pyxis.engineering.redhat.com/v1/images?filter=top_layer_id=="
    image_id = list_layer_imageid[i]
    team_url = base_url + str(image_id)
    # print(team_url)
    r = requests.get(team_url, auth=kerberos_auth, verify=False)
    data = json.loads(r.content)
    if len(data["data"]) > 0:
        if len(data["data"][0]["parsed_data"]["labels"]) > 0:
            df = pd.DataFrame(data["data"][0]["parsed_data"]["labels"])
            table = pd.pivot_table(
                df, values="value", aggfunc=lambda x: x, columns="name"
            )
            table["image_id"] = list_layer_imageid[i]
            table = table.set_index("image_id")
            dataframe = dataframe.append(table)

100%|██████████| 752/752 [11:26<00:00,  1.10it/s]


In [16]:
df_image_layerid = dataframe
df_image_layerid.head()

architecture  \
image_id                                                          
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...       x86_64   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...       x86_64   
sha256:7ca359fa9862549d7254cc7410b8f5e44c54412e...       x86_64   
sha256:d497c3fc656f819799726c642c2c36405ea113e8...       x86_64   
sha256:e365f51ada9c737b4f2a0e588991d4aa9e24af66...       x86_64   

                                                                    build-date  \
image_id                                                                         
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...  2021-02-10T00:15:26.710361   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...  2021-03-30T18:30:01.465792   
sha256:7ca359fa9862549d7254cc7410b8f5e44c54412e...  2020-12-08T19:08:54.854539   
sha256:d497c3fc656f819799726c642c2c36405ea113e8...  2020-12-16T00:02:32.230857   
sha256:e365f51ada9c737b4f2a0e588991d4aa9e24af66...  2021-01-04T17:40:28.142615   

                                                                         com.redhat.build-host  \
image_id                                                                                         
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...  cpt-1004.osbs.prod.upshift.rdu2.redhat.com   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...  cpt-1003.osbs.prod.upshift.rdu2.redhat.com   
sha256:7ca359fa9862549d7254cc7410b8f5e44c54412e...  cpt-1001.osbs.prod.upshift.rdu2.redhat.com   
sha256:d497c3fc656f819799726c642c2c36405ea113e8...  cpt-1006.osbs.prod.upshift.rdu2.redhat.com   
sha256:e365f51ada9c737b4f2a0e588991d4aa9e24af66...  cpt-1007.osbs.prod.upshift.rdu2.redhat.com   

                                                     com.redhat.component  \
image_id                                                                    
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...         ubi8-container   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...         ubi8-container   
sha256:7ca359fa9862549d7254cc7410b8f5e44c54412e...         ubi7-container   
sha256:d497c3fc656f819799726c642c2c36405ea113e8...     s2i-core-container   
sha256:e365f51ada9c737b4f2a0e588991d4aa9e24af66...  rh-nodejs14-container   

                                                                             com.redhat.license_terms  \
image_id                                                                                                
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...  https://www.redhat.com/en/about/red-hat-end-us...   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...  https://www.redhat.com/en/about/red-hat-end-us...   
sha256:7ca359fa9862549d7254cc7410b8f5e44c54412e...  https://www.redhat.com/en/about/red-hat-end-us...   
sha256:d497c3fc656f819799726c642c2c36405ea113e8...  https://www.redhat.com/en/about/red-hat-end-us...   
sha256:e365f51ada9c737b4f2a0e588991d4aa9e24af66...  https://www.redhat.com/en/about/red-hat-end-us...   

                                                                                          description  \
image_id                                                                                                
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...  The Universal Base Image is designed and engin...   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...  The Universal Base Image is designed and engin...   
sha256:7ca359fa9862549d7254cc7410b8f5e44c54412e...  The Universal Base Image is designed and engin...   
sha256:d497c3fc656f819799726c642c2c36405ea113e8...  The s2i-core image provides any images layered...   
sha256:e365f51ada9c737b4f2a0e588991d4aa9e24af66...  Node.js 14 available as container is a base pl...   

                                                   distribution-scope  \
image_id                                                                
sha256:fdb393d8227cbe9756537d3f215a3098ae797bd4...             public   
sha256:64607cc74f9cbe0e12f167547df0cf661de5a8b1...             public   
sha256:7ca359fa

In [18]:
df_image_layerid.shape

(143, 68)

We were able to create a dataframe which maps 143 sha's of the image layer id provided in the image layer dataset out of 752 sha's in the image layer dataset. These mapped datasets are then stored in the DH_PLAYPEN bucket.

In [19]:
# Uploading the mapping dataset in the DH-PLAYPEN bucket
parquet_buffer = io.BytesIO()
df_image_layerid.to_parquet(parquet_buffer)
s3_obj = s3.Object(
    s3_bucket, "ccx/fingerprinting/image_layers/df_image_layerid.parquet"
)
status = s3_obj.put(Body=parquet_buffer.getvalue())

---

# Mapping the SHA's in`image id` column of Containers dataset

In this section, we will map the SHA's in the containers dataset to their product name, summary, vendor, and other attributes. We will first form a list of unique image_id's from the containers dataset. Using that list, we will be doing the web scraping followed by the formation of the dataframe with image_id and corresponding attributes.

In [192]:
# Listing out the SHA's of image_id
arr_cont_imageid = containers_df.image_id.unique()
list_cont_imageid = arr_cont_imageid.tolist()

In [193]:
dataframe = pd.DataFrame([])

for i in tqdm(range(len(list_cont_imageid))):
    base_url = "https://pyxis.engineering.redhat.com/v1/images?filter=image_id=="
    image_id = list_cont_imageid[i]
    team_url = base_url + str(image_id)
    # print(team_url)
    r = requests.get(team_url, auth=kerberos_auth, verify=False)
    data = json.loads(r.content)
    if len(data["data"]) > 0:
        if len(data["data"][0]["parsed_data"]["labels"]) > 0:
            df = pd.DataFrame(data["data"][0]["parsed_data"]["labels"])
            table = pd.pivot_table(
                df, values="value", aggfunc=lambda x: x, columns="name"
            )
            table["image_id"] = list_cont_imageid[i]
            table = table.set_index("image_id")
            dataframe = dataframe.append(table)

100%|██████████| 5477/5477 [1:25:56<00:00,  1.06it/s]  


In [194]:
df_cont_image_id = dataframe
df_cont_image_id.head()

License architecture  \
image_id                                                                   
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...  ASL 2.0       x86_64   
sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf3...   GPLv2+       x86_64   
sha256:a693c315b775c693dc49c19b7f217762676bc287...   GPLv2+       x86_64   
sha256:d9c64d038f16e04c52142bc9e7dfa0645ce7e348...   GPLv2+       x86_64   
sha256:af32a5e5489c79461bf64c6e55bd34e1c1c1bd94...   GPLv2+       x86_64   

                                                                    build-date  \
image_id                                                                         
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...  2021-05-04T21:54:58.392948   
sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf3...  2021-05-04T22:05:48.018566   
sha256:a693c315b775c693dc49c19b7f217762676bc287...  2021-05-04T21:53:09.591552   
sha256:d9c64d038f16e04c52142bc9e7dfa0645ce7e348...  2021-05-04T22:19:44.056847   
sha256:af32a5e5489c79461bf64c6e55bd34e1c1c1bd94...  2021-05-04T22:18:44.217005   

                                                                         com.redhat.build-host  \
image_id                                                                                         
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...  cpt-1007.osbs.prod.upshift.rdu2.redhat.com   
sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf3...  cpt-1006.osbs.prod.upshift.rdu2.redhat.com   
sha256:a693c315b775c693dc49c19b7f217762676bc287...  cpt-1006.osbs.prod.upshift.rdu2.redhat.com   
sha256:d9c64d038f16e04c52142bc9e7dfa0645ce7e348...  cpt-1003.osbs.prod.upshift.rdu2.redhat.com   
sha256:af32a5e5489c79461bf64c6e55bd34e1c1c1bd94...  cpt-1004.osbs.prod.upshift.rdu2.redhat.com   

                                                                         com.redhat.component  \
image_id                                                                                        
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...                  kube-rbac-proxy-container   
sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf3...  ose-multus-admission-controller-container   
sha256:a693c315b775c693dc49c19b7f217762676bc287...                       multus-cni-container   
sha256:d9c64d038f16e04c52142bc9e7dfa0645ce7e348...            ose-egress-router-cni-container   
sha256:af32a5e5489c79461bf64c6e55bd34e1c1c1bd94...  ose-containernetworking-plugins-container   

                                                             com.redhat.license_terms  \
image_id                                                                                
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...  https://www.redhat.com/agreements   
sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf3...  https://www.redhat.com/agreements   
sha256:a693c315b775c693dc49c19b7f217762676bc287...  https://www.redhat.com/agreements   
sha256:d9c64d038f16e04c52142bc9e7dfa0645ce7e348...  https://www.redhat.com/agreements   
sha256:af32a5e5489c79461bf64c6e55bd34e1c1c1bd94...  https://www.redhat.com/agreements   

                                                                                          description  \
image_id                                                                                                
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...  This is a proxy, that can perform Kubernetes R...   
sha256:edb9aaacf421c6dc45b20324e8699cec02f26bf3...  This is a component of OpenShift Container Pla...   
sha256:a693c315b775c693dc49c19b7f217762676bc287...  This is a component of OpenShift Container Pla...   
sha256:d9c64d038f16e04c52142bc9e7dfa0645ce7e348...                       CNI Plugin for Egress Router   
sha256:af32a5e5489c79461bf64c6e55bd34e1c1c1bd94...  This is a component of OpenShift Container Pla...   

                                                   distribution-scope  \
image_id                                                                
sha256:f46f210d6023bec16e68340b484a8881ce46d5e6...             public  

In [197]:
df_cont_image_id.shape

(2478, 113)

Here, we successfully did the mapping for 2478 sha's of the image_id for container dataset out of 5477 sha's of image_id.

In [198]:
# Uploading the mapping dataset in the DH-PLAYPEN bucket
parquet_buffer = io.BytesIO()
df_cont_image_id.to_parquet(parquet_buffer)
s3_obj = s3.Object(s3_bucket, "ccx/fingerprinting/containers/df_cont_image_id.parquet")
status = s3_obj.put(Body=parquet_buffer.getvalue())

The corresponding mapped dataframe is saved in DH-PLAYPEN bucket.

---

# Conclusion

The notebook does takes some time to run. In the notebook, we were able to map the product name with the corresponding image_id from the image layer dataset and the container dataset. They mapped dataframe are then saved in the DH-PLAYPEN bucket. 

As next steps, in forthcoming notebooks we will integrate the dataframe generated here with the sample insight operator workload dataframe, and perfrom EDA in context of the combined information.